Detectors/TRD/workflow/include/TRDWorkflow/GainCalibSpec.h

```diff
@@ -0,0 +1,154 @@
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#ifndef O2_TRD_GAINCALIBSPEC_H
+#define O2_TRD_GAINCALIBSPEC_H
+
+/// \file   GainCalibSpec.h
+/// \brief DPL device for steering the TRD gain time slot based calibration
+/// \author Ole Schmidt, Gauthier Legras
+
+#include "TRDCalibration/CalibratorGain.h"
+#include "DetectorsCalibration/Utils.h"
+#include "DataFormatsTRD/GainCalibHistos.h"
+#include "CommonUtils/MemFileHelper.h"
+#include "Framework/Task.h"
+#include "Framework/ConfigParamRegistry.h"
+#include "Framework/ControlService.h"
+#include "Framework/WorkflowSpec.h"
+#include "CCDB/CcdbApi.h"
+#include "CCDB/CcdbObjectInfo.h"
+#include "Framework/CCDBParamSpec.h"
+#include "DetectorsBase/GRPGeomHelper.h"
+#include <chrono>
+
+using namespace o2::framework;
+
+namespace o2
+{
+namespace calibration
+{
+
+class GainCalibDevice : public o2::framework::Task
+{
+ public:
+  GainCalibDevice(std::shared_ptr<o2::base::GRPGeomRequest> req) : mCCDBRequest(req) {}
+  void init(o2::framework::InitContext& ic) final
+  {
+    o2::base::GRPGeomHelper::instance().setRequest(mCCDBRequest);
+    auto enableOutput = ic.options().get<bool>("enable-root-output");
+    auto slotL = ic.options().get<uint32_t>("sec-per-slot");
+    auto delay = ic.options().get<uint32_t>("max-delay");
+    mCalibrator = std::make_unique<o2::trd::CalibratorGain>(enableOutput);
+    mCalibrator->setSlotLengthInSeconds(slotL);
+    mCalibrator->setMaxSlotsDelay(delay);
+    mCalibrator->createOutputFile(); // internally checks if output is enabled
+  }
+
+  void finaliseCCDB(o2::framework::ConcreteDataMatcher& matcher, void* obj) final
+  {
+    o2::base::GRPGeomHelper::instance().finaliseCCDB(matcher, obj);
+  }
+
+  void run(o2::framework::ProcessingContext& pc) final
+  {
+    auto runStartTime = std::chrono::high_resolution_clock::now();
+    o2::base::GRPGeomHelper::instance().checkUpdates(pc);
+    mCalibrator->retrievePrev(pc);
+    auto dataGainCalib = pc.inputs().get<o2::trd::GainCalibHistos>("input");
+    o2::base::TFIDInfoHelper::fillTFIDInfo(pc, mCalibrator->getCurrentTFInfo());
+    LOG(info) << "Processing TF " << mCalibrator->getCurrentTFInfo().tfCounter << " with " << dataGainCalib.getNEntries() << " GainCalibHistos entries";
+    mCalibrator->process(dataGainCalib);
+    sendOutput(pc.outputs());
+    std::chrono::duration<double, std::milli> runDuration = std::chrono::high_resolution_clock::now() - runStartTime;
+    LOGP(info, "Duration for run method: {} ms", std::chrono::duration_cast<std::chrono::milliseconds>(runDuration).count());
+  }
+
+  void endOfStream(o2::framework::EndOfStreamContext& ec) final
+  {
+    LOG(info) << "Finalizing calibration";
+    mCalibrator->checkSlotsToFinalize(o2::calibration::INFINITE_TF);
+    mCalibrator->closeOutputFile();
+    sendOutput(ec.outputs());
+  }
+
+  void stop() final
+  {
+    mCalibrator->closeOutputFile();
+  }
+
```

Comment:
This mechanism has changed upstream since you last looked, for reference see #11155.